diff --git a/dist/cpexcel.js b/dist/cpexcel.js
index f821c7a56fab98ec4bc54624d2e15a4e7a922422..cbe7cbdd37257198c06df955a8b31c920d42f0c0 100644
--- a/dist/cpexcel.js
+++ b/dist/cpexcel.js
@@ -804,7 +804,7 @@ if (typeof module !== 'undefined' && module.exports) module.exports = cptable;
   "use strict";
   if(typeof cptable === "undefined") {
     if(typeof require !== "undefined"){
-      var cpt = require('./cpt' + 'able');
+      var cpt = cptable;
       if (typeof module !== 'undefined' && module.exports) module.exports = factory(cpt);
       else root.cptable = factory(cpt);
     } else throw new Error("cptable not found");
diff --git a/ods.js b/ods.js
index 9d498b0e7f1c4335a12a1e297d8b203045127763..ab005f1521879763607811218bc24b9a12cde740 100644
--- a/ods.js
+++ b/ods.js
@@ -7,7 +7,7 @@ var ODS = {};
 var get_utils = function() {
 	if(typeof XLSX !== 'undefined') return XLSX.utils;
 	if(typeof module !== "undefined" && typeof require !== 'undefined') try {
-		return require('../' + 'xlsx').utils;
+		return require('./' + 'xlsx').utils;
 	} catch(e) {
 		try { return require('./' + 'xlsx').utils; }
 		catch(ee) { return require('xl' + 'sx').utils; }
diff --git a/xlsx.js b/xlsx.js
index 15b58b1f53ef0c4b6a3b7fa53dfe2bfe51c33f4d..7bb85df151ea7f8e2cf0434358af67e5fac14a4a 100644
--- a/xlsx.js
+++ b/xlsx.js
@@ -7810,21 +7810,54 @@ return function parse_ws_xml_data(sdata, s, opts, guess) {
 	}
 }; })();
 
+/* lpj 改变行高 */
+var DEF_PPI = 96, PPI = DEF_PPI;
+function px2pt(px) { return px * 96 / PPI; }
+function pt2px(pt) { return pt * PPI / 96; }
 function write_ws_xml_data(ws, opts, idx, wb) {
-	var o = [], r = [], range = safe_decode_range(ws['!ref']), cell, ref, rr = "", cols = [], R, C;
-	for(C = range.s.c; C <= range.e.c; ++C) cols[C] = encode_col(C);
-	for(R = range.s.r; R <= range.e.r; ++R) {
-		r = [];
-		rr = encode_row(R);
-		for(C = range.s.c; C <= range.e.c; ++C) {
-			ref = cols[C] + rr;
-			if(ws[ref] === undefined) continue;
-			if((cell = write_ws_xml_cell(ws[ref], ref, ws, opts, idx, wb)) != null) r.push(cell);
-		}
-		if(r.length > 0) o[o.length] = (writextag('row', r.join(""), {r:rr}));
-	}
-	return o.join("");
+    var o = [], r = [], range = safe_decode_range(ws['!ref']), cell="", ref, rr = "", cols = [], R=0, C=0, rows = ws['!rows'];
+    var dense = Array.isArray(ws);
+    var params = ({r:rr}), row, height = -1;
+    for(C = range.s.c; C <= range.e.c; ++C) cols[C] = encode_col(C);
+    for(R = range.s.r; R <= range.e.r; ++R) {
+        r = [];
+        rr = encode_row(R);
+        for(C = range.s.c; C <= range.e.c; ++C) {
+            ref = cols[C] + rr;
+            var _cell = dense ? (ws[R]||[])[C]: ws[ref];
+            if(_cell === undefined) continue;
+            if((cell = write_ws_xml_cell(_cell, ref, ws, opts, idx, wb)) != null) r.push(cell);
+        }
+        if(r.length > 0 || (rows && rows[R])) {
+            params = ({r:rr});
+            if(rows && rows[R]) {
+                row = rows[R];
+                if(row.hidden) params.hidden = 1;
+                height = -1;
+                if (row.hpx) height = px2pt(row.hpx);
+                else if (row.hpt) height = row.hpt;
+                if (height > -1) { params.ht = height; params.customHeight = 1; }
+                if (row.level) { params.outlineLevel = row.level; }
+            }
+            o[o.length] = (writextag('row', r.join(""), params));
+        }
+    }
+    if(rows) for(; R < rows.length; ++R) {
+        if(rows && rows[R]) {
+            params = ({r:R+1});
+            row = rows[R];
+            if(row.hidden) params.hidden = 1;
+            height = -1;
+            if (row.hpx) height = px2pt(row.hpx);
+            else if (row.hpt) height = row.hpt;
+            if (height > -1) { params.ht = height; params.customHeight = 1; }
+            if (row.level) { params.outlineLevel = row.level; }
+            o[o.length] = (writextag('row', "", params));
+        }
+    }
+    return o.join("");
 }
+/* lpj 改变行高 end */
 
 var WS_XML_ROOT = writextag('worksheet', null, {
 	'xmlns': XMLNS.main[0],